File: src/Terrain.cpp
#include "../include/Terrain.h"
#include <iostream>
#include <random>
#include <iomanip>
#include <algorithm>
#include <cmath>

// ANSI color codes for green terminal output
#define RESET   "\033[0m"
#define GREEN   "\033[32m"
#define BRIGHT_GREEN "\033[1;32m"

Terrain::Terrain(int w, int h) : width(w), height(h) {
    grid.resize(height, std::vector<TerrainType>(width, TerrainType::NORMAL));
    elevationMap.resize(height, std::vector<double>(width, 0.0));
    windResistance.resize(height, std::vector<double>(width, 0.0));
}

void Terrain::setTerrain(int x, int y, TerrainType type) {
    if (isValidPosition(Point(x, y))) {
        grid[y][x] = type;
    }
}

TerrainType Terrain::getTerrain(int x, int y) const {
    if (isValidPosition(Point(x, y))) {
        return grid[y][x];
    }
    return TerrainType::OBSTACLE;
}

void Terrain::setElevation(int x, int y, double elevation) {
    if (isValidPosition(Point(x, y))) {
        elevationMap[y][x] = elevation;
    }
}

double Terrain::getElevation(int x, int y) const {
    if (isValidPosition(Point(x, y))) {
        return elevationMap[y][x];
    }
    return 0.0;
}

void Terrain::setWindResistance(int x, int y, double resistance) {
    if (isValidPosition(Point(x, y))) {
        windResistance[y][x] = resistance;
    }
}

double Terrain::getWindResistance(int x, int y) const {
    if (isValidPosition(Point(x, y))) {
        return windResistance[y][x];
    }
    return 0.0;
}

bool Terrain::isValidPosition(const Point& pos) const {
    return pos.x >= 0 && pos.x < width && pos.y >= 0 && pos.y < height;
}

bool Terrain::isObstacle(const Point& pos) const {
    if (!isValidPosition(pos)) return true;
    return grid[pos.y][pos.x] == TerrainType::OBSTACLE;
}

bool Terrain::isPassable(const Point& pos) const {
    return isValidPosition(pos) && !isObstacle(pos);
}

double Terrain::getMovementCost(const Point& pos) const {
    if (!isValidPosition(pos)) return OBSTACLE_COST;
    
    double baseCost = NORMAL_COST;
    TerrainType type = grid[pos.y][pos.x];
    
    switch (type) {
        case TerrainType::NORMAL:
            baseCost = NORMAL_COST;
            break;
        case TerrainType::HILL:
            baseCost = HILL_COST;
            break;
        case TerrainType::WIND_ZONE:
            baseCost = WIND_COST;
            break;
        case TerrainType::OBSTACLE:
            return OBSTACLE_COST;
        default:
            baseCost = NORMAL_COST;
    }
    
    // Add elevation cost
    double elevationCost = elevationMap[pos.y][pos.x] * 0.5;
    
    // Add wind resistance
    double windCost = windResistance[pos.y][pos.x] * 0.3;
    
    return baseCost + elevationCost + windCost;
}

double Terrain::getHeuristicCost(const Point& from, const Point& to) const {
    // Manhattan distance with diagonal movement consideration
    double dx = std::abs(to.x - from.x);
    double dy = std::abs(to.y - from.y);
    
    // Euclidean distance for more accurate heuristic
    return std::sqrt(dx * dx + dy * dy);
}

void Terrain::visualizeTerrain() const {
    std::cout << "=== Terrain Map ===\n";
    std::cout << "  ";
    for (int x = 0; x < width; x++) {
        std::cout << std::setw(2) << x;
    }
    std::cout << "\n";
    
    for (int y = 0; y < height; y++) {
        std::cout << std::setw(2) << y;
        for (int x = 0; x < width; x++) {
            std::cout << " " << getTerrainChar(grid[y][x]);
        }
        std::cout << "\n";
    }
    std::cout << "==================\n";
}

void Terrain::visualizePath(const std::vector<Point>& path) const {
    std::cout << GREEN << "  ";
    for (int x = 0; x < width; x++) {
        std::cout << std::setw(2) << x;
    }
    std::cout << RESET << "\n";
    
    for (int y = 0; y < height; y++) {
        std::cout << GREEN << std::setw(2) << y << RESET;
        for (int x = 0; x < width; x++) {
            Point current(x, y);
            bool isOnPath = false;
            
            // Check if this position is on the path
            for (const Point& pathPoint : path) {
                if (pathPoint == current) {
                    isOnPath = true;
                    break;
                }
            }
            
            char displayChar;
            if (isOnPath) {
                // Check if it's start or end
                if (current == path.front()) {
                    displayChar = 'S';
                } else if (current == path.back()) {
                    displayChar = 'D';
                } else {
                    displayChar = '*';  // Use * instead of Unicode character
                }
                std::cout << " " << BRIGHT_GREEN << displayChar << RESET;
            } else {
                displayChar = getTerrainChar(grid[y][x]);
                std::cout << " " << GREEN << displayChar << RESET;
            }
        }
        std::cout << "\n";
    }
}

char Terrain::getTerrainChar(TerrainType type) const {
    switch (type) {
        case TerrainType::NORMAL: return '.';
        case TerrainType::HILL: return '^';
        case TerrainType::OBSTACLE: return 'O';
        case TerrainType::WIND_ZONE: return 'W';
        case TerrainType::START: return 'S';
        case TerrainType::END: return 'D';
        default: return '?';
    }
}

std::vector<Point> Terrain::getNeighbors(const Point& pos) const {
    std::vector<Point> neighbors;
    
    // 8-directional movement (including diagonals)
    const int dx[] = {-1, -1, -1, 0, 0, 1, 1, 1};
    const int dy[] = {-1, 0, 1, -1, 1, -1, 0, 1};
    
    for (int i = 0; i < 8; i++) {
        Point neighbor(pos.x + dx[i], pos.y + dy[i]);
        if (isPassable(neighbor)) {
            neighbors.push_back(neighbor);
        }
    }
    
    return neighbors;
}

void Terrain::generateRandomTerrain(double obstacleProb, double hillProb, double windProb) {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_real_distribution<> dis(0.0, 1.0);
    
    for (int y = 0; y < height; y++) {
        for (int x = 0; x < width; x++) {
            double roll = dis(gen);
            
            if (roll < obstacleProb) {
                setTerrain(x, y, TerrainType::OBSTACLE);
            } else if (roll < obstacleProb + hillProb) {
                setTerrain(x, y, TerrainType::HILL);
                setElevation(x, y, dis(gen) * 5.0); // Random elevation 0-5
            } else if (roll < obstacleProb + hillProb + windProb) {
                setTerrain(x, y, TerrainType::WIND_ZONE);
                setWindResistance(x, y, dis(gen) * 3.0); // Random wind resistance 0-3
            } else {
                setTerrain(x, y, TerrainType::NORMAL);
                setElevation(x, y, dis(gen) * 2.0); // Small elevation variation
            }
        }
    }
}

void Terrain::addObstacle(const Point& pos) {
    setTerrain(pos.x, pos.y, TerrainType::OBSTACLE);
}

void Terrain::addHill(const Point& pos) {
    setTerrain(pos.x, pos.y, TerrainType::HILL);
    setElevation(pos.x, pos.y, 3.0);
}

void Terrain::addWindZone(const Point& pos) {
    setTerrain(pos.x, pos.y, TerrainType::WIND_ZONE);
    setWindResistance(pos.x, pos.y, 2.0);
}

File: src/terrain.rs
use std::collections::HashMap;

// ANSI color codes for green terminal output
const RESET: &str = "\x1b[0m";
const GREEN: &str = "\x1b[32m";
const BRIGHT_GREEN: &str = "\x1b[1;32m";

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Point {
    pub x: i32,
    pub y: i32,
}

impl Point {
    pub fn new(x: i32, y: i32) -> Self {
        Point { x, y }
    }
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum TerrainType {
    Normal,
    Hill,
    Obstacle,
    WindZone,
    Start,
    End,
}

pub struct Terrain {
    grid: Vec<Vec<TerrainType>>,
    elevation_map: Vec<Vec<f64>>,
    wind_resistance: Vec<Vec<f64>>,
    width: usize,
    height: usize,
}

impl Terrain {
    pub fn new(width: usize, height: usize) -> Self {
        Terrain {
            grid: vec![vec![TerrainType::Normal; width]; height],
            elevation_map: vec![vec![0.0; width]; height],
            wind_resistance: vec![vec![0.0; width]; height],
            width,
            height,
        }
    }

    pub fn set_terrain(&mut self, x: usize, y: usize, terrain_type: TerrainType) {
        if self.is_valid_position(&Point::new(x as i32, y as i32)) {
            self.grid[y][x] = terrain_type;
        }
    }

    pub fn get_terrain(&self, x: usize, y: usize) -> TerrainType {
        if self.is_valid_position(&Point::new(x as i32, y as i32)) {
            self.grid[y][x]
        } else {
            TerrainType::Obstacle
        }
    }

    pub fn set_elevation(&mut self, x: usize, y: usize, elevation: f64) {
        if self.is_valid_position(&Point::new(x as i32, y as i32)) {
            self.elevation_map[y][x] = elevation;
        }
    }

    pub fn get_elevation(&self, x: usize, y: usize) -> f64 {
        if self.is_valid_position(&Point::new(x as i32, y as i32)) {
            self.elevation_map[y][x]
        } else {
            0.0
        }
    }

    pub fn set_wind_resistance(&mut self, x: usize, y: usize, resistance: f64) {
        if self.is_valid_position(&Point::new(x as i32, y as i32)) {
            self.wind_resistance[y][x] = resistance;
        }
    }

    pub fn get_wind_resistance(&self, x: usize, y: usize) -> f64 {
        if self.is_valid_position(&Point::new(x as i32, y as i32)) {
            self.wind_resistance[y][x]
        } else {
            0.0
        }
    }

    pub fn is_valid_position(&self, pos: &Point) -> bool {
        pos.x >= 0 && pos.x < self.width as i32 && pos.y >= 0 && pos.y < self.height as i32
    }

    pub fn is_obstacle(&self, pos: &Point) -> bool {
        if !self.is_valid_position(pos) {
            return true;
        }
        matches!(self.grid[pos.y as usize][pos.x as usize], TerrainType::Obstacle)
    }

    pub fn is_passable(&self, pos: &Point) -> bool {
        self.is_valid_position(pos) && !self.is_obstacle(pos)
    }

    pub fn get_movement_cost(&self, pos: &Point) -> f64 {
        if !self.is_valid_position(pos) {
            return 1000.0;
        }

        let x = pos.x as usize;
        let y = pos.y as usize;
        
        let base_cost = match self.grid[y][x] {
            TerrainType::Normal => 1.0,
            TerrainType::Hill => 3.0,
            TerrainType::WindZone => 2.0,
            TerrainType::Obstacle => 1000.0,
            _ => 1.0,
        };

        if base_cost >= 1000.0 {
            return base_cost;
        }

        // Add elevation cost
        let elevation_cost = self.elevation_map[y][x] * 0.5;
        
        // Add wind resistance
        let wind_cost = self.wind_resistance[y][x] * 0.3;
        
        base_cost + elevation_cost + wind_cost
    }

    pub fn get_heuristic_cost(&self, from: &Point, to: &Point) -> f64 {
        let dx = (to.x - from.x) as f64;
        let dy = (to.y - from.y) as f64;
        (dx * dx + dy * dy).sqrt()
    }

    pub fn get_neighbors(&self, pos: &Point) -> Vec<Point> {
        let mut neighbors = Vec::new();
        
        // 8-directional movement (including diagonals)
        let directions = [
            (-1, -1), (-1, 0), (-1, 1),
            (0, -1),           (0, 1),
            (1, -1),  (1, 0),  (1, 1)
        ];
        
        for (dx, dy) in directions {
            let neighbor = Point::new(pos.x + dx, pos.y + dy);
            if self.is_passable(&neighbor) {
                neighbors.push(neighbor);
            }
        }
        
        neighbors
    }

    pub fn visualize_path(&self, path: &Vec<Point>) {
        // Create a HashMap for quick path lookup
        let path_set: HashMap<Point, usize> = path.iter().enumerate().map(|(i, &p)| (p, i)).collect();
        
        // Print column headers
        print!("{}  {}", GREEN, RESET);
        for x in 0..self.width {
            print!("{}{:2}{}", GREEN, x, RESET);
        }
        println!();
        
        // Print grid with path
        for y in 0..self.height {
            print!("{}{:2}{}", GREEN, y, RESET);
            for x in 0..self.width {
                let current = Point::new(x as i32, y as i32);
                
                if let Some(&path_index) = path_set.get(&current) {
                    // This position is on the path
                    let display_char = if path_index == 0 {
                        'S'
                    } else if path_index == path.len() - 1 {
                        'D'
                    } else {
                        '*'
                    };
                    print!(" {}{}{}", BRIGHT_GREEN, display_char, RESET);
                } else {
                    // Regular terrain
                    let terrain_char = self.get_terrain_char(self.grid[y][x]);
                    print!(" {}{}{}", GREEN, terrain_char, RESET);
                }
            }
            println!();
        }
    }

    fn get_terrain_char(&self, terrain_type: TerrainType) -> char {
        match terrain_type {
            TerrainType::Normal => '.',
            TerrainType::Hill => '^',
            TerrainType::Obstacle => 'O',
            TerrainType::WindZone => 'W',
            TerrainType::Start => 'S',
            TerrainType::End => 'D',
        }
    }

    pub fn get_width(&self) -> usize {
        self.width
    }

    pub fn get_height(&self) -> usize {
        self.height
    }
}